### PR TITLE
Make a hard reset of named barrier mutexes on signal

### PR DESCRIPTION
**Pull request by oxidase**
# Issue

For #3477 added a simple signal handler to cleanup the shared barrier mutexes. It will not prevent stale locks after OOM kills. `set_terminate` is not really needed, because SIGABRT will be sent to the process by default `terminate` handler.

## Tasklist
 - [x] review
 - [x] adjust for comments

## Requirements / Relations
 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

**Comment by daniel-j-h** [comment]
The overall approach looks good - some question re. signals, as I don't grasp the edge cases around them fully:
- What happens if a signal comes in while our lock-release signal handler is running?
- `signal` is conforming to the C standard so _should_ be portable - can we test this on macOS and Windows before merging
- Should the signal handler be `[[noreturn]]` `noexcept`?
- Should the signal handler call `_Exit` to shut down hard? (\*)

(\*)



https://github.com/adishavit/Terminators

**Comment by oxidase** [comment]
- it is implementation dependent: signal handler can reset to DFL or block other signals, the best would be to use `sigaction` where behavior can be tuned, but it is platform-dependent. 
- yes, `signal` is defined in §18.10
- noreturn  with _exit is ok, noexcept is also ok, but here https://github.com/boostorg/interprocess/blob/develop/include/boost/interprocess/sync/named_sharable_mutex.hpp#L140 just a comment :)
- good point! i will add "_exit(128 + signum);" and noreturn

File: src/tools/store.cpp [hunk 1]
```diff
@@ -87,8 +90,22 @@ bool generateDataStoreOptions(const int argc,
     return true;
 }
 
+[[ noreturn ]] void CleanupSharedBarriers(int signum)
+{ // Here the lock state of named mutexes is unknown, make a hard cleanup
+    osrm::storage::SharedBarriers::resetCurrentRegions();
+    osrm::storage::SharedBarriers::resetRegions1();
+    osrm::storage::SharedBarriers::resetRegions2();
+    _exit(128 + signum);
```
Comment (oxidase) [inline src/tools/store.cpp]: :+1: updated

**Comment by daniel-j-h** [comment]
Looks good - let's wait for Appveyor / Travis and then this should be ready to merge. Would maybe love to test this in action, at least n macOS. Maybe @danpat can you try to force this behavior.

File: src/tools/store.cpp [hunk 2]
```diff
@@ -87,8 +91,22 @@ bool generateDataStoreOptions(const int argc,
     return true;
 }
 
+[[ noreturn ]] void CleanupSharedBarriers(int signum)
+{ // Here the lock state of named mutexes is unknown, make a hard cleanup
+    osrm::storage::SharedBarriers::resetCurrentRegions();
+    osrm::storage::SharedBarriers::resetRegions1();
```
Comment (TheMarex) [inline src/tools/store.cpp]: This is potentially dangerous:

These locks govern the shared mutexes for the data region. They should only be reset if osrm-datastore actually had a lock on them.

If these locks are just removed while a libosrm instance is running this can lead to crashes on the next data update or exceptions because locks will try to unlock mutexes that are in an unlocked state.

Comment (oxidase) [inline src/tools/store.cpp]: right, this is dangerous, but reaching this point with a custom or default handler is a bad sign for the whole IPC because locks will be in unknown state. I will remove reseting regions, so it can be controlled by `--max-wait` and reset at https://github.com/Project-OSRM/osrm-backend/blob/master/src/storage/storage.cpp#L158-L164
